### DISCUSSION TITLE
Add support for Regex script

### DISCUSSION THREAD

**Discussion by arthur-fontaine** [pnpm/pnpm#16078] [Ideas]
### Contribution

- [X] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

As a user of an "external script runner", it can be very verbose to create a new script.

For example, if you are using [Wireit](https://github.com/google/wireit), you first have to add a key-value to the `scripts` object in the `package.json`, then you can write your script. Same problem with Turbo.

### Describe the solution you'd like

In the same way that pnpm allows Regex to be used in the `run` command, we could add support for "Regex as a key" in the `scripts` object in the package.json.

For example:

```json
{
  ...,
  "scripts": {
    "/.*/": "echo \"I run for every command\""
  }
}
```

Running `pnpm run a` or `pnpm run b` or any other script will always echo "I run for every command".

> [!NOTE]
> If there is both a Regex script and a literal script, both scripts will run.

---

We could also add support for substitution.

For example:

```json
{
  ...,
  "scripts": {
    "/.*/": "turbo run \\0"
  }
}
```

Running `pnpm run a` will run `turbo run a`

> [!IMPORTANT]
> I use "\\digit" to match the pattern as `sed` does. There may be a better pattern to use.

### Describe the drawbacks of your solution

pnpm already allows the use of Regex in the `run` command (https://pnpm.io/cli/run#running-multiple-scripts). These features may overlap.

If "Regex in scripts" is implemented, I think that using a Regex in the `run` command should match the source of the regex.

Example:

```json
{
  ...,
  "scripts": {
    "/.*/": "echo \"I run for every command\"",
    "test:a": "echo \"test a\"",
    "test:b": "echo \"test b\""
  }
}
```

_Result:_
| Command | Result |
| --- | --- |
| `pnpm run "/.*/"` | *run all scripts* |
| `pnpm run "/test:*/"` | *run `test:a` and `test:b`* |
| `pnpm run "/\/\.\*\//"` | *run `/.*/`* |

### Describe alternatives you've considered

I have found no alternative to this feature.

**Comment by KSXGitHub** [comment]
There is a point where it is better to use a proper taskrunner (such as a `make` or `just`) than to rely on the package manager to do everything that isn't related to managing packages.

Anyway, I don't think those should be in `scripts`. Maybe `regexScripts` or `pnpm.regexScripts`?

**Comment by arthur-fontaine** [comment]
Why do you think it should not be in `scripts`?
`regexScripts` is not in the [package.json specs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) and in my opinion we should avoid breaking them as much as possible.
Using `scripts` will not break the package.json specs. pnpm will just behave differently than npm. Moreover, I think that this feature could be optional and be enabled through the `.npmrc` (just like `enable-pre-post-scripts`) (`enable-regex-scripts`?) to prevent breaks even more.

**Comment by KSXGitHub** [comment]
> Why do you think it should not be in `scripts`?

Because `scripts` aren't supposed to be regex.

> `regexScripts` is not in the [package.json specs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) and in my opinion we should avoid breaking them as much as possible.

There's also `pnpm.regexScripts`.

Anyway, your use-case is to redirect all `pnpm run` to another taskrunner, right? How about `pnpm.taskRunner` which would be invoked if `pnpm run` does not find matching script name in `scripts`?

**Comment by zkochan** [comment]
I agree that this is out of scope for pnpm. We concentrate on package management not running tasks. If you need such complex tasks configuration, maybe you should use a dedicated tool or create your own and use it instead of `pnpm run`.

**Comment by everett1992** [comment]
This doesn't seem compatible with pnpm's regex to run multiple tasks - if the regex is '/test:.*/' and turbo has test:a, test:b, test:c pnpm could not enumerate a, b, c. 

why aren't you calling turbo directly, why do you need to enter thru pn and package.json scripts?

**Comment by arthur-fontaine** [comment]
Except I didn't understand it well, your example is one of the case where this feature would be useful.

Instead of having to declare 3 PNPM scripts to run the turbo's test:a, test:b and test:c, you can declare only one test:.*.

It would be something like

~~~
"/test:(.*)/": "turbo run test:\0"
~~~

Instead of

~~~
"test:a": "turbo run test:a",
"test:b": "turbo run test:b",
"test:c": "turbo run test:c",
~~~

The feature is to improve experience working with the monorepo tools or tools having script running feature, while keeping `pnpm run` central to the developer experience (instead of having some tasks running with `pnpm run`, others with `turbo run`, etc.)

**Comment by everett1992** [comment]
How does pnpm know to run test:a, test:b, test:c and not test:1, test:2,
test:3?

**Comment by arthur-fontaine** [comment]
Because the user typed `pnpm run test:a`, which match the `"/test:(.*)/"` PNPM script, which run `turbo run test:\0`, where the `\0` is replaced by the regex param (`(.*)`)

To add precisions, this feature does not replace the existing support for running scripts with regex. It cannot execute multiple scripts on its own. Instead, it works the reversed way: it can match multiple run arguments using a regex.
